syz-manager/filter.go

```diff
@@ -0,0 +1,246 @@
+)
+
+type CoverFilter struct {
+	enableFilter    bool
+	kcovFilterStart uint32
+	kcovFilterSize  uint32
+	kcovFilterEnd   uint32
+	weightedPCs     map[uint32]float32
+
+	pcsBitmapPath      string
+	targetLittleEndian bool
+	targetArch         string
+}
+
+func (mgr *Manager) initKcovFilter() {
+	covFilterConfig := make(map[string][]string)
+	if err := json.Unmarshal(mgr.cfg.CovFilter, &covFilterConfig); err != nil {
+		log.Logf(0, "no coverage filter is enabled")
+		mgr.kcovFilter.enableFilter = false
+		return
+	}
+	files := covFilterConfig["files"]
+	funcs := covFilterConfig["functions"]
+	rawPCs := covFilterConfig["pcs"]
+	if len(files) == 0 && len(funcs) == 0 && len(rawPCs) == 0 {
+		mgr.kcovFilter.enableFilter = false
+		return
+	}
+	mgr.kcovFilter.enableFilter = true
+	log.Logf(0, "initialize coverage information...")
+	if err := initCover(mgr.sysTarget, mgr.cfg.KernelObj, mgr.cfg.KernelSrc, mgr.cfg.KernelBuildSrc); err != nil {
+		log.Logf(0, "failed to generate coverage profile: %v", err)
+		log.Fatalf("coverage filter cannot be initialized without coverage profile")
+	}
+
+	mgr.kcovFilter.targetLittleEndian = mgr.sysTarget.LittleEndian
+	mgr.kcovFilter.targetArch = mgr.sysTarget.Arch
+	mgr.kcovFilter.pcsBitmapPath = mgr.cfg.Workdir + "/" + "syz-cover-bitmap"
+	mgr.kcovFilter.initWeightedPCs(files, funcs, rawPCs)
+}
+
+func (mgr *Manager) getWeightedPCs() bool {
+	return mgr.kcovFilter.enableFilter
+}
+
+func (filter *CoverFilter) initWeightedPCs(files, functions, rawPCsFiles []string) {
+	filter.weightedPCs = make(map[uint32]float32)
+	filesRegexp := getFileRegexp(files)
+	funcsRegexp := getFuncRegexp(functions)
+
+	enabledFiles := make(map[string]bool)
+	enabledFuncs := make(map[string]bool)
+	for _, f := range rawPCsFiles {
+		rawFile, err := os.Open(f)
+		if err != nil {
+			log.Logf(0, "failed to open raw PCs file: %v", err)
+		}
+		for {
+			var encode uint64
+			_, err := fmt.Fscanf(rawFile, "0x%x\n", &encode)
+			if err != nil {
+				break
+			}
+			pc := uint32(encode & 0xffffffff)
+			weight := float32((encode >> 32) & 0xffff)
+			filter.weightedPCs[pc] = weight
+		}
+		rawFile.Close()
+	}
+	pcs := reportGenerator.PCs()
+	for _, e := range pcs {
+		frame := e[len(e)-1]
+		fullpc := cover.NextInstructionPC(filter.targetArch, frame.PC)
+		pc := uint32(fullpc & 0xffffffff)
+		for _, r := range funcsRegexp {
+			if ok := r.MatchString(frame.Func); ok {
+				enabledFuncs[frame.Func] = true
+				filter.weightedPCs[pc] = 1.0
+			}
+		}
+		for _, r := range filesRegexp {
+			if ok := r.MatchString(frame.File); ok {
+				enabledFiles[frame.File] = true
+				enabledFuncs[frame.Func] = true
+				filter.weightedPCs[pc] = 1.0
+			}
+		}
+		if _, ok := filter.weightedPCs[pc]; ok {
+			enabledFuncs[frame.Func] = true
+		}
+	}
+	for f := range enabledFuncs {
+		log.Logf(1, "enabled func: %s", f)
+	}
+	for f := range enabledFiles {
+		log.Logf(1, "enabled file: %s", f)
+	}
+}
+
+func getFileRegexp(files []string) []regexp.Regexp {
```

Comment:
getFileRegexp and getFuncRegexp add a bit extended regexp-like syntax.
The syntax is somehow confusing and non-consistent, e.g. `*` in the middle has one meaning and syntax (needs to be preceded with `.` like in regexps), but at end it has different meaning and syntax (like in wildcards). `^` is appended in some cases, but not in others.
There are also no tests for negative cases (incorrect syntax).
If we want a full-fledged new specification language, this will need more work.
But also it does not add any significant of value, everything it's doing can be as easily expressed with normal regexps, with 0 additional code/tests on our side and with syntax/semantics already familiar for users. E.g. "net/dccp/**" can be replaced with "^net/dccp/", or maybe I would not even bother with "^" and do just "net/dccp" because if it matches anything new that's probably interesting files as well. So the syntax is even simpler.
So I would drop all of this code/tests and just let users specify regexps.